File: src/main/java/com/mohe/spring/repository/UserRepository.java
package com.mohe.spring.repository;

import com.mohe.spring.entity.User;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.time.OffsetDateTime;
import java.util.Optional;

@Repository
public interface UserRepository extends JpaRepository<User, Long> {
    
    Optional<User> findByEmail(String email);
    
    Optional<User> findByNickname(String nickname);
    
    boolean existsByEmail(String email);
    
    boolean existsByNickname(String nickname);
    
    @Modifying
    @Query("UPDATE User u SET u.lastLoginAt = :loginTime WHERE u.id = :userId")
    void updateLastLoginAt(@Param("userId") Long userId, @Param("loginTime") OffsetDateTime loginTime);
    
    @Modifying
    @Query("UPDATE User u SET u.isOnboardingCompleted = :completed WHERE u.id = :userId")
    void updateOnboardingCompleted(@Param("userId") Long userId, @Param("completed") boolean completed);
}

File: src/main/java/com/mohe/spring/repository/UserRepository.kt
package com.mohe.spring.repository

import com.mohe.spring.entity.User
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.time.OffsetDateTime
import java.util.*

@Repository
interface UserRepository : JpaRepository<User, Long> {
    
    fun findByEmail(email: String): Optional<User>
    
    fun findByNickname(nickname: String): Optional<User>
    
    fun existsByEmail(email: String): Boolean
    
    fun existsByNickname(nickname: String): Boolean
    
    @Modifying
    @Query("UPDATE User u SET u.lastLoginAt = :loginTime WHERE u.id = :userId")
    fun updateLastLoginAt(@Param("userId") userId: Long, @Param("loginTime") loginTime: OffsetDateTime)
    
    @Modifying
    @Query("UPDATE User u SET u.isOnboardingCompleted = :completed WHERE u.id = :userId")
    fun updateOnboardingCompleted(@Param("userId") userId: Long, @Param("completed") completed: Boolean)
}